x/gov/keeper/deposit.go

```diff
@@ -162,6 +163,76 @@ func (keeper Keeper) AddDeposit(ctx sdk.Context, proposalID uint64, depositorAdd
 	return activatedVotingPeriod, nil
 }
 
+// chargeDeposit will charge proposal cancellation fee (deposits * proposal_cancel_burn_rate)  and
+// send to a destAddress if defined or burn otherwise.
+// Remaining funds are send back to the depositor.
+func (keeper Keeper) chargeDeposit(ctx sdk.Context, proposalID uint64, destAddress, proposalCancelRate string, totalDeposits []sdk.Coin) error {
+	store := ctx.KVStore(keeper.storeKey)
+
+	rate := sdk.MustNewDecFromStr(proposalCancelRate)
+	var cancellationCharges sdk.Coins
+
+	if rate.IsPositive() {
+		for _, deposit := range totalDeposits {
+			burnAmount := sdk.NewCoin(
+				deposit.Denom,
+				sdk.NewDecFromInt(deposit.Amount).Mul(rate).RoundInt(),
+			)
+			cancellationCharges = cancellationCharges.Add(burnAmount)
+		}
+	}
+
+	// burn the cancellation fee or sent the cancellation charges to destination address.
+	if !cancellationCharges.IsZero() {
+		// get the distribution module account address
+		distributionAddress := keeper.authKeeper.GetModuleAddress(disttypes.ModuleName)
+		if distributionAddress.String() == destAddress {
+			err := keeper.distrkeeper.FundCommunityPool(ctx, cancellationCharges, keeper.ModuleAccountAddress())
+			if err != nil {
+				return err
+			}
+		} else {
+			if len(destAddress) == 0 {
+				// burn the remaining deposits also
+				err := keeper.bankKeeper.BurnCoins(ctx, types.ModuleName, cancellationCharges)
+				if err != nil {
+					return err
+				}
+			} else {
+				destAccAddress := sdk.MustAccAddressFromBech32(destAddress)
+				err := keeper.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, destAccAddress, cancellationCharges)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	// proposal cancel rate is not 100% then we have remaining deposits after the cancellation charges
+	if !rate.Equal(sdk.MustNewDecFromStr("1")) {
+		// send remaining (deposits - cancellationCharges) to depositors
+		for _, deposits := range keeper.GetDeposits(ctx, proposalID) {
+			depositerAddress := sdk.MustAccAddressFromBech32(deposits.Depositor)
+			var remainingAmount sdk.Coins
+			for _, deposit := range deposits.Amount {
+				remainAmount := sdk.NewCoin(
+					deposit.Denom,
+					sdk.NewDecFromInt(deposit.Amount).Mul(sdk.NewDec(1).Sub(rate)).RoundInt(),
```

Comment:
I think `RoundInt()` is dangerous, here and L179.

Imagine the scenario:
- rate is 0.5
- proposer submits with deposit = 1stake
- no other depositors

Then:
- `cancelCharge = round(1*0.5) = 1`
- `remainAmt = round(1*(1-0.5)) = 1`

and we break the invariant: cancelCharge + remainAmt = totalDeposit.

Reply:
@robert-zaremba @alexanderbez  any suggestions how to handle this error ,because sdk.Coin or sdk.DecCoin required amount as `math.Int`

Reply:
It looks like the approach in this PR has changed significantly since I last reviewed it. If we need to maintain the invariant `cancelCharge + remainAmt = totalDeposit`, then I would just perform multiplication on a single variable and subtract the remainder. Meaning, `remainAmt = totalDeposit - cancelCharger`.

Reply:
@gsk967 could you add a test to cover this scenario. @alexanderbez suggestion is good.

Reply:
https://github.com/cosmos/cosmos-sdk/pull/13010/commits/a5a2b57ac90fb3117b1a3c99189513ae5e5460e9#diff-72b620cb17246319713aafae5b0e10536e2ea6d5aefbe107ffc7982362b1a514R374-R383

Reply:
Test is still not added afaict